### DISCUSSION TITLE
Nextjs and GraphQL

### DISCUSSION THREAD

**Discussion by marceloverdijk** [vercel/next.js#35241] [Help]
I'm wondering if it is easy to include some (embedded) GraphQL data layer as Gatsby has?

The rationale is I have a big json file (or alternatively a sqlite database) and I want to use that to build a SSG website.
So from `getStaticPaths` and `getStaticPropos` I would like to call an embedded GraphQL endpoint to provide the needed data.

Using GraphQL would allow me to tune the business schema to my own needs by adding resolvers etc.

As I'm using SSG the GraphQL should not be part of the generated code.

Anybody experience with such a setup?

**Comment by icyJoseph** [comment]
Let me see if I understand your scope.

I have an application that does `getStaticProps` from the github API (which is, of course, not hosted by me). To do this, I simply post to their endpoint with my queries.

To create my queries, I do something like:

```js
 const github = await queryGitHub( GET_USER, { login: "icyJoseph" })
```

Where `GET_USER` is something like:

```js
import { print } from "graphql";
import gql from "graphql-tag";

const ItemShowcaseFragment = gql`
  fragment showcaseItem on Repository {
    name
  }
`;

const GET_USER_DOC = gql`
  query getUser($login: String!, $from: DateTime, $to: DateTime) {
    user(login: $login) {
      bio
      company
      location
      login
      name
      avatarUrl
    }
  }
  ${ItemShowcaseFragment}
`;

const GET_YEAR_CONTRIBUTIONS_DOC = gql`
  query getYearContributions($login: String!, $from: DateTime, $to: DateTime) {
    user(login: $login) {
      contributionsCollection(from: $from, to: $to) {
        ...contributions
      }
    }
  }
  ${ContributionsFragment}
`;

export const GET_USER = print(GET_USER_DOC);
```

I also have an `api/route` which can receive parametrised calls that go out to the GitHub GraphQL API. This way, no graphQL handling code is ever shipped to my clients.

If you are thinking that your Next.js instance should have the `GraphQL` endpoint, I have to warn you that, that's not necessarily possible.

At build time, there's no lambdas, in fact, there's no running server, so even if you had `api/graphl`, such endpoint will be non-existent. 

This is why, `URLs` ought to be absolute when using `fetch` inside `getStaticProps` and `getServerSideProps`. Note that even `axios` won't help you (Next.js polyfills fetch for node), it'll just not find the endpoint needed. 

So in short, the idea of having  an `api/graphql` endpoint exposed by your application to render on demand is not so easily done, and it is not possible at build time. You'd need to have access on `getServerSideProps` to the absolute URL of your application, any relative paths will just fail.

You could however, on `getStaticProps` or `getServeSideProps` connect to the external resource, or use `cwd` like so https://nextjs.org/docs/api-reference/data-fetching/get-static-props#reading-files-use-processcwd, to read from the file system.

```js
// or getStaticProps
export const getServerSideProps = async () => {
  const res = await fs.readFile(
    path.resolve(process.cwd(), "public", "waves.min.svg"),
    "utf-8"
  );

  return { props: { } };
}
```

To send a GraphQL request, it's just a `POST` request, with the query and variables as body payload.

I don't know if this answers your question, but perhaps it helps you get going?

**Reply by marceloverdijk** [reply]
Thx @icyJoseph for your reply.

Basically I would like to use some "temporary" embedded GraphQL server during the Nextjs build process.
(Similar as Gatsby but just for some relational model of my own; no need to have full exposure of all Nextjs intern types/nodes).

If I would run the build process locally I could fire up a separate GraphQL server (e.g. Apollo Client using Prisma) but I want to the building using Github Action, so firing up a separate GraphQL server process is not an option as far as I know.

One of the benefits using GraphQL to driver my data needs - e.g. compared to using Prisma for direct db queries - is that the GraphQL plays nicely with JSON as output. E.g. when using Prisma.io I face a lot of issues with date types and decimals (as not natively supported by json).
And I could even use graphql-codegen to generate typesafe GraphQL types (again fully compatible with JSON hydration) that I could use in getStaticProps.

To give the bigger picture what I'm trying to achieve: I'm generating a static website of ~5K pages (various types) which source is either a json file or relational (sqlite) database.

Alternative is to use the sqlite database as source and pre-generate a json for each page with all data needed.
But one drawback is that it less typesafe than e.g. using a GraphQL schema.

**Reply by icyJoseph** [reply]
What about MSW? You could set it up to catch requests going out to, say, `localhost:3001`, even though there's no real server there, and then it is up to you to write your own resolvers. https://mswjs.io/docs/api/graphql for more information, and here for a repo using it,  https://github.com/vercel/next.js/tree/canary/examples/with-msw.

Again, at build time you enable the MSW `require` and initiate the server, then it will catch requests going out to a given endpoint. There you take over, and write your own resolvers.

**Reply by marceloverdijk** [reply]
Interesting, I wasn't aware of MSW, need to check it out, thx!
